ls1intum/Artemis
# Support for Checkstyle and PMD

### Checklist

### Motivation and Context
This PR adapts Artemis to handle the changes introduced by https://github.com/ls1intum/bamboo-server-notification-plugin/pull/9. 

### Description
_Server:_
- Add dependencies for Checkstyle and PMD to pom.xml
- Define student working directory `assignment/src` for programming exercises as a constant and use this value for cutting off CI specific paths and setting the `<sourceDirectory>` in the pom.xml
- Store static code analysis issues as Json in `feedback.detailText`. Makes it easier to parse the data in the client.
- Adapt tests

_Client:_
- Adapt `result-details.component` to handle the new static code analysis feedback format
- Fix bug where fetching of build logs in `result-details.component` could prevent the displaying the details if 403 was returned
- Adapt annotations to new feedback format
- Adapt tests

### Steps for Testing


_Normal programming exercise_
1. Create a new programming exercise **without** static code analysis
2. Participate in the exercise
3. Verify that results are created correctly

_Programming exercise with static code analysis_
1. Create a new programming exercise **with** static code analysis
2. Participate in the exercise
3. Verify that results are created correctly and that static code analysis feedback is available (code editor annotations, result details)

### Test Coverage

![FeedbackServiceCoverage](https://user-images.githubusercontent.com/16407766/91910734-dbd7b400-ecaf-11ea-8278-e821a2ef8221.png)
![build-output-coverage](https://user-images.githubusercontent.com/16407766/94380144-984c5a80-0134-11eb-89db-2641257dd0d9.png)
![locationTextPipeCoverage](https://user-images.githubusercontent.com/16407766/94380145-984c5a80-0134-11eb-80ec-ac4fc1b72c28.png)
![result-detail-coverage](https://user-images.githubusercontent.com/16407766/94380146-98e4f100-0134-11eb-94cb-3b6ed83c93f9.png)
